enhancements/network/ovn-kubernetes-localnet-api.md

```diff
@@ -0,0 +1,558 @@
+type ClusterUserDefinedNetworkSpec struct {
+    ...
+    // +required
+    Network NetworkSpec `json:"network"`
+}
+```
+
+As of today the CUDN has validation to ensure `spec.network` is immutable:
+```go
+// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Network spec is immutable"
+```
+This validation should be changed in a way it allows mutating localnet topology configurations, 
+at least MTU, VLAN, physicalNetworkName and excludeSubnets. 
+The proposed change is to invert the immutable-spec validation on CUDN `spec.network` as follows:
+1. Remove mutation restrictions on `spec.network`.
+2. Add CEL rule validations to restrict mutations of:
+   `spec.network.topology`, `spec.network.layer2`, and `spec.network.layer3`.
+3. Add CEL rule validations to restrict mutation of localnet topology settings: 
+   `spec.network.localnet.role`, `spec.network.localnet.subnets` and `spec.network.topology.ipam`.
+
+##### Localnet topology configuration type
+Introduce new topology configuration type for localnet - `LocalnetConfig`.
+The CUDN CRD `spec.network` should feature proposed localnet topology configuration.
+
+The Layer2 and Layer3 configuration types (`Layer2Config` & `Layer3Config`) `role` field is defined with `NetworkRole` type.
+The `NetworkRole` type has the following enum validation, allowing `Secondary` and `Primary` values:
+```
+// +kubebuilder:validation:Enum=Primary;Secondary
+```
+The proposed localnet config type (`LocalnetConfig`) `role` field would have to accept `Secondary` value only.
+In order to avoid misleading CRD scheme, the enum validation should be moved closer to each `NetworkRole` usage:
+1. Remove the enum validation exist on `NetworkRole` definition
+2. At the `Layer2Config` definition, add enum validation to `role` field allowing: `Primary` or `Secondary`.
+3. At the `Layer3Config` definition, add enum validation to `role` field allowing: `Primary` or `Secondary`.
+4. At the proposed `LocalnetConfig` definition, add enum validation to `role` field allowing `Secondary` only.
+
+```go
+// +kubebuilder:validation:XValidation:rule="has(self.ipam) && has(self.ipam.mode) && self.ipam.mode != 'Enabled' || has(self.subnets)", message="Subnets is required with ipam.mode is Enabled or unset"
+// +kubebuilder:validation:XValidation:rule="!has(self.ipam) || !has(self.ipam.mode) || self.ipam.mode != 'Disabled' || !has(self.subnets)", message="Subnets must be unset when ipam.mode is Disabled"
+// +kubebuilder:validation:XValidation:rule="!has(self.excludeSubnets) || has(self.excludeSubnets) && has(self.subnets)", message="excludeSubnets must be unset when subents is unset"
+// +kubebuilder:validation:XValidation:rule="!has(self.subnets) || !has(self.mtu) || !self.subnets.exists_one(i, isCIDR(i) && cidr(i).ip().family() == 6) || self.mtu >= 1280", message="MTU should be greater than or equal to 1280 when IPv6 subent is used"
```

Comment:
Exists one checks that there is exactly one item that matches the condition, you here are looking for at least one, in which case you should use `exists`. It is computationally cheaper

Reply:
The subents field expects the following options: 
1. Single IPv4 CIDR
2. Single IPv6 CIDR 
3. Single IPv4 CIDR and single IPv6 CIDR
In case an IPv6 CIDR present (option 1 or 2) it should ensure MTU is greater then 1280.
Since there should be single item of IPv6 familty we use the `exists_one` operation.
This is true for the existing topologies we have today (Layer2 & Layer3).
Make sense?

Reply:
Ok great, then you don't need `exists_one` here, since you're already checking there's only one IPv6 address